File: hashTable.ts
import { hashCode } from "./hashCode";
import { getIndexFromHashcCode } from "./getIndexFromHashCode";
import { LinkedList, LinkedListNode } from "./linkedList";

export class HashTable {
  arr: LinkedList[] = new Array(10);

  // put in hashtable
  put(key: string, value: any): void {
    const keyHashCode: number = hashCode(key);
    const indexFromHashCode: number = getIndexFromHashcCode(
      keyHashCode,
      this.arr
    );

    if (!this.arr[indexFromHashCode]) {
      this.arr[indexFromHashCode] = new LinkedList([key, value]);
    } else {
      let currentLinkedList: LinkedList = this.arr[indexFromHashCode];
      let current: LinkedListNode = currentLinkedList.head;
      let found: boolean = false;
      if (current.data[0] == key) {
        current.data[1] = value;
        found = true;
        return;
      }
      while (current.next) {
        if (current.next.data[0] === key) {
          found = true;
          current.next.data[1] = value;
          return;
        }
        current = current.next;
      }
      if (found === false) {
        currentLinkedList.append([key, value]);
      }
    }
  }

  // get value from hsashtable by key
  get(key: string): any {
    const keyHashCode = hashCode(key);
    const indexFromHashCode: number = getIndexFromHashcCode(
      keyHashCode,
      this.arr
    );
    let linkedListAtIndex = this.arr[indexFromHashCode];
    if (!linkedListAtIndex) {
      return null;
    }
    let found: boolean = false;
    let current: LinkedListNode = linkedListAtIndex.head;
    if (current.data[0] === key) {
      found = true;
      return current.data[1];
    } else {
      while (current.next) {
        if (current.next.data[0] === key) {
          found = true;
          return current.next.data[1];
        } else {
          current = current.next;
        }
      }
    }
    if (found === false) {
      return null;
    }
  }
}

const MyHashTable = new HashTable();
MyHashTable.put("mohamad1", 2010);
MyHashTable.put("mohamad", 1996);
MyHashTable.put("mohamaddd", 2018);
MyHashTable.put("mohamaddd", 202000);
MyHashTable.put("mohamadddd", 2018);

File: hashTable.py
from linkedList import LinkedList
from getIndexFromHashCode import getIndexFromHashcode
from hashCode import generateHashCode


class HashTable:
    array = [None]*10

    def put(self, key, value):
        keyHashCode = generateHashCode(key)
        index = getIndexFromHashcode(keyHashCode, self.array)
        if self.array[index] == None:
            self.array[index] = LinkedList([key, value])
        elif self.array[index] != None and type(self.array[index] is LinkedList):
            current = self.array[index].head
            found = False
            while(current):
                if current.value[0] == key:
                    current.value[1] = value
                    found = True
                    break
                current = current.next
            if found == False:
                self.array[index].append([key, value])

    def get(self, key):
        keyHashCode = generateHashCode(key)
        index = getIndexFromHashcode(keyHashCode, self.array)
        if(self.array[index] == None):
            return None
        currentLinkedList = self.array[index]
        current = currentLinkedList.head
        while(current):
            if (current.value[0] == key):
                return current.value[1]
            current = current.next

    def getEntries(self):
        entries = []
        for i in range(0, len(self.array)):
            if(self.array[i] != None):
                currentLinkedList = self.array[i]
                current = currentLinkedList.head
                while(current):
                    entries.append(
                        {"key": current.value[0], "value": current.value[1]})
                    current = current.next
        return entries
